pipeline_dp/pipeline_operations.py

```diff
@@ -75,6 +87,51 @@ def sample_fixed_per_key(self, col, n: int, stage_name: str):
   def count_per_element(self, col, stage_name: str):
     return col | stage_name >> combiners.Count.PerElement()
 
+
+class SparkRDDOperations(PipelineOperations):
+  """Apache Spark RDD adapter."""
+
+  def map(self, rdd, fn, stage_name: str = None):
+    return rdd.map(fn)
+
+  def map_tuple(self, rdd, fn, stage_name: str = None):
+    return rdd.map(fn)
+
+  def map_values(self, rdd, fn, stage_name: str = None):
+    return rdd.mapValues(fn)
+
+  def group_by_key(self, rdd, stage_name: str = None):
+    """Group the values for each key in the RDD into a single sequence.
+
+    Args:
+      rdd: input RDD
+      stage_name: not used
+
+    Returns:
+      An RDD of tuples in which the type of the second item
+      is the pyspark.resultiterable.ResultIterable.
+
+    """
+    return rdd.groupByKey()
+
+  def filter(self, rdd, fn, stage_name: str = None):
+    return rdd.filter(fn)
+
+  def keys(self, rdd, stage_name: str = None):
+    return rdd.keys()
+
+  def values(self, rdd, stage_name: str = None):
+    return rdd.values()
+
+  def sample_fixed_per_key(self, rdd, n: int, stage_name: str = None):
+    return rdd.mapValues(lambda x: [x])\
```

Comment:
thanks! I like this implementation, it's simple, scalable and provides uniform sampling

Reply:
Thanks for drawing attention to uniformity! In fact, in general `reduceByKey` won't give us uniform sampling. An RDD in Spark consists of partitions, and [`reduceByKey` firstly reduces each partition before sending the results for final reductions](https://spark.apache.org/docs/latest/api/scala/org/apache/spark/rdd/PairRDDFunctions.html#reduceByKey(func:(V,V)=%3EV):org.apache.spark.rdd.RDD[(K,V)]). It's implemented this way to reduce shuffling (i.e. traffic between executors/workers). So for a given key the less values reside in a specific partition, the more likely (when comparing with values for this key from other partitions) these values will get to the final sample.

Should we keep this implementation or use the previous one for now?

Ideally, we'd have to collect statistics about cardinality for each key to the driver in the first phase and based on this to sample the corresponding number of values from each partition. But it's complicated and more time is needed for this implementation.

Reply:
Ah, correct, this is not uniform sampling. The uniformness is not a strict requirement. 

Could you please add a comment that this sampling is not uniform? It's important to inform about such caveats :). I think scalability is more important that uniformness here.

Reply:
Added the comment :)